cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/WindowsAggregationDefinition.java

```diff
@@ -0,0 +1,179 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+
+/**
+ * Specifies how a window operation should be executed.
+ */
+public class WindowsAggregationDefinition extends AggregationDefinition {
+  private WindowsAggregationDefinition(List<Expression> windowsExpressions,
+                                         Map<String, Expression> selectExpressions) {
+        super(windowsExpressions, selectExpressions);
+  }
+    /**
+     * @return A builder to create a WindowsAggregationDefinition.
+     */
+  public static Builder builder() {
+        return new Builder();
+    }
+
+  public enum FilterFunction {
+    RANK,
+    DENSE_RANK,
+    PERCENT_RANK,
+    N_TILE,
+    ROW_NUMBER,
+    MEDIAN,
+    CONTINUOUS_PERCENTILE,
+    DISCRETE_PERCENTILE,
+    LEAD,
+    LAG,
+    FIRST,
+    LAST,
+    CUMULATIVE_DISTRIBUTION,
+    ACCUMULATE;
+  }
+
+  public static class FilterExpression {
+    private final Expression expression;
+    private final FilterFunction filterFunction;
+
+    /**
+     * Creates a new {@link FilterExpression} using the specified {@link Expression} and {@link FilterFunction}.
+     * @param expression the expression using which filtering of duplicate records is to be performed
+     * @param filterFunction the function to be applied on the duplicate records to select a single record
+     */
+    public FilterExpression(Expression expression, FilterFunction filterFunction) {
+        this.expression = expression;
+        this.filterFunction = filterFunction;
+    }
+
+    /**
+     *
+     * @return the expression used for filtering duplicate records
+     */
+    public Expression getExpression() {
+        return expression;
+    }
+
+        /**
+         *
+         * @return the function applied to select a single record from duplicate records
+         */
+        public FilterFunction getFilterFunction() {
+            return filterFunction;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            FilterExpression that = (FilterExpression) o;
+            return Objects.equals(getExpression(), that.getExpression())
+              && getFilterFunction() == that.getFilterFunction();
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(getExpression(), getFilterFunction());
+        }
+    }
+    /**
+     * Builds a WindowsAggregationDefinition using fields to group by and fields to select.
+     * The fields to select must be specified, while the fields to group by are optional.
+     */
+    public static class Builder {
+        private List<Expression> windowsExpressions;
+        private Map<String, Expression> selectExpressions;
+
+        public Builder() {
+            windowsExpressions = Collections.emptyList();
```

Comment:
I don't think these are mutable instances. You need to initialize them using new ArrayList() and new HashMap();